src/webcore/promise.rs

```diff
@@ -0,0 +1,252 @@
+use std;
+use std::error::Error as _Error;
+use std::marker::PhantomData;
+use webcore::once::Once;
+use webcore::value::{Value, Reference, ConversionError};
+use webcore::try_from::{TryInto, TryFrom};
+use web::error::Error;
+use futures::{Future, Poll, Async};
+use futures::unsync::oneshot::{Receiver, channel};
+use webcore::promise_executor::spawn;
+
+
+/// A `Promise` object represents the eventual completion (or failure) of an asynchronous operation, and its resulting value.
+///
+/// In most situations you shouldn't use this, use [`PromiseFuture`](struct.PromiseFuture.html) instead.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)
+// https://www.ecma-international.org/ecma-262/6.0/#sec-promise-objects
+pub struct Promise( Reference );
+
+reference_boilerplate! {
+    Promise,
+    instanceof Promise
+}
+
+impl Promise {
+    /// This function should rarely be needed, use [`PromiseFuture`](struct.PromiseFuture.html) instead.
+    ///
+    /// This function is used for two different purposes:
+    ///
+    /// 1. If you have a JavaScript value which is not a `Promise` but you want to wrap it in a `Promise`, you can use `Promise::promisify(value)`.
+    ///    In this situation, it is recommended to use [`futures::future::ok`](https://docs.rs/futures/0.1.18/futures/future/fn.ok.html) instead.
+    ///
+    /// 2. If you have a JavaScript value which is a Promise-like object (it has a `then` method) but it isn't a true `Promise`, you can use
+    ///    `Promise::promisify(value)` to convert it into a true `Promise`. This situation is rare, but it can happen if you are using a Promise
+    ///    library such as jQuery or Bluebird.
+    ///
+    /// # Examples
+    ///
+    /// Convert a JavaScript value to a `Promise`:
+    ///
+    /// ```rust
+    /// Promise::promisify(js!( return 5; ))
+    /// ```
+    ///
+    /// Convert a Promise-like object to a `Promise`:
+    ///
+    /// ```rust
+    /// // jQuery Promise
+    /// Promise::promisify(js!( return $.get("test.php"); ))
+    ///
+    /// // Bluebird Promise
+    /// Promise::promisify(js!( return bluebird_promise.timeout(1000); ))
+    /// ```
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve)
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promise.resolve
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promise-resolve-functions
+    // https://www.ecma-international.org/ecma-262/6.0/#sec-promiseresolvethenablejob
+    pub fn promisify( input: Value ) -> Promise {
```

Comment:
Why not just keep the name `resolve`?

Reply:
Because that name is super confusing.

If I see the name `promisify` I have a pretty good idea of what it does (it probably converts something into a Promise).

If I see the name `resolve`, I would never in a million years think that it converts something into a Promise. I would probably assume that it is somehow mutating an existing Promise or something (which isn't what it does).

Also, I have considered making it so that the `promisify` function only works on fake Promises, not arbitrary values, so its implementation won't be as simple as just `Promise.resolve(value)`

I'm open for suggestions on changing its name, but `resolve` is just a bad name for what it does. It's a bad name in JavaScript and an even worse name in Rust.

Reply:
There is value in sticking to the original method names though, and I don't think `promisify` is any more useful a name - you can figure that much out by the type signature.

As a compromise, how about `Promise::resolved(..)` - this at least better indicates that it's constructing a new, resolved promise rather than modifying an existing one, while being very close to the original name.

Reply:
I can compromise with `Promise::resolved`, though I do worry about the confusion for people who are used to JavaScript (they need to remember that it ends in `d` in Rust, but doesn't in JavaScript)

I think for that reason `Promise::resolve` would be better than `Promise::resolved`, though both are quite bad.

The more I think about it, the more I prefer the solution of changing `Promise::promisify` to only convert fake Promises, and then the name makes more sense. Or perhaps rename it to `Promise::convert`, I'm still open to name suggestions.